packages/react/src/components/TreeView/TreeNode.js

```diff
@@ -0,0 +1,276 @@
+  const nodesWithProps = React.Children.map(children, (node) => {
+    if (React.isValidElement(node)) {
+      return React.cloneElement(node, {
+        active,
+        depth: depth + 1,
+        disabled,
+        onTreeSelect,
+        selected,
+        tabIndex: (!node.props.disabled && -1) || null,
+      });
+    }
+  });
+  const isActive = active === id;
+  const isSelected = selected.includes(id);
+  const treeNodeClasses = classNames(className, `${prefix}--tree-node`, {
+    [`${prefix}--tree-node--active`]: isActive,
+    [`${prefix}--tree-node--disabled`]: disabled,
+    [`${prefix}--tree-node--selected`]: isSelected,
+    [`${prefix}--tree-node--with-icon`]: Icon,
+    [`${prefix}--tree-leaf-node`]: !children,
+    [`${prefix}--tree-parent-node`]: children,
+  });
+  const toggleClasses = classNames(`${prefix}--tree-parent-node__toggle-icon`, {
+    [`${prefix}--tree-parent-node__toggle-icon--expanded`]: expanded,
+  });
+  const handleToggleClick = (event) => {
+    if (onToggle) {
+      onToggle(event, { id, isExpanded: !expanded, label, value });
+    }
+    setExpanded(!expanded);
+  };
+  const handleClick = (event) => {
+    event.stopPropagation();
+    if (!disabled) {
+      if (onTreeSelect) {
+        onTreeSelect(event, { id, label, value });
+      }
+      if (onNodeSelect) {
+        onNodeSelect(event, { id, label, value });
+      }
+      if (rest.onClick) {
+        rest.onClick(event);
+      }
+    }
+  };
+  const handleKeyDown = (event) => {
+    if (matches(event, [keys.ArrowLeft, keys.ArrowRight, keys.Enter])) {
+      event.stopPropagation();
+    }
+    if (match(event, keys.ArrowLeft)) {
+      const findParentTreeNode = (node) => {
+        if (node.classList.contains(`${prefix}--tree-parent-node`)) {
+          return node;
+        }
+        if (node.classList.contains(`${prefix}--tree`)) {
+          return null;
+        }
+        return findParentTreeNode(node.parentNode);
+      };
+      if (children && expanded) {
+        onToggle(event, { id, isExpanded: false, label, value });
+        setExpanded(false);
+      } else {
+        /**
+         * When focus is on a leaf node or a closed parent node, move focus to
+         * its parent node (unless its depth is level 1)
+         */
+        findParentTreeNode(currentNode.current.parentNode)?.focus();
+      }
+    }
+    if (children && match(event, keys.ArrowRight)) {
+      if (expanded) {
+        /**
+         * When focus is on an expanded parent node, move focus to the first
+         * child node
+         */
+        currentNode.current.lastChild.firstChild.focus();
+      } else {
+        onToggle(event, { id, isExpanded: true, label, value });
+        setExpanded(true);
+      }
+    }
+    if (matches(event, [keys.Enter, keys.Space])) {
+      event.preventDefault();
+      handleClick(event);
+    }
+    if (rest.onKeyDown) {
+      rest.onKeyDown(event);
+    }
+  };
+
+  useEffect(() => {
```

Comment:
Would it help if we had a version of: https://github.com/reach/reach-ui/blob/main/packages/utils/src/index.tsx#L60 so we could use a layout effect here?

Reply:
can you explain the reasoning behind your question a bit more? I wasn't intending to use layout effect here so not sure if I would benefit from a helper like this here

Reply:
@emyarod it seemed like the effect was computing layout/positioning so thought it would be appropriate 👀

Reply:
I see, I was under the impression that `useEffect` would be more preferable unless there is some visual lag associated with the effect